### ISSUE TITLE
Creating new repository empty and without user with permissions?

### ISSUE THREAD

**Issue by rehmsen** [ericpaulbishop/redmine_git_hosting#67]
Not sure whether this is the right place to ask this, but there does not seem a forum/mailing list, and no one ever answers over at redmine.org.

When I create a new repository for some project, the gitolite.conf will have an additonal line declaring the project, but there will not be any user having read or write access to it. How am I adding write permissions to some user? When I go to Project Settings->Repository->Users, I see a message "No Data to Display" which seems to be because the log of the repository is scanned for users, but since the repository is brand new, there are no commits so no users are found. When I go to Project -> Repository, I am told to push my stuff to the newly created repository, which I cannot as I do not have write access to it. Am I overlooking something? How do I give some initial user permissions?

**Comment by perryk** [comment]
The user will need to have a public key set and be a member of the project which has the repo.

After that their name and permission should show up in the gitolite.conf file.

**Comment by ericpaulbishop** [comment]
This should now be fixed in v0.4.0

**Comment by rehmsen** [comment]
I pulled 0.4.0, adjusted the configuration as described in the updated README and ran plugin migrations:
They fail with the following console output:

```
rm: cannot remove `/tmp/redmine_git_hosting/gitolite-admin': Permission denied
[sudo] password for username: 
error: could not lock config file repositories/rep1.git/config: No such file or directory
error: could not lock config file repositories/rep1.git/config: No such file or directory
error: could not lock config file repositories/rep2.git/config: No such file or directory
error: could not lock config file repositories/rep2.git/config: No such file or directory
error: could not lock config file repositories/rep3.git/config: No such file or directory
error: could not lock config file repositories/rep3.git/config: No such file or directory
rm: cannot remove `/tmp/redmine_git_hosting/gitolite-admin': Permission denied
```

and later 
    ==  SetMirrorRolePermissions: migrating =======================================
    rake aborted!
    An error has occurred, all later migrations canceled:

```
undefined method `add_permission!' for nil:NilClass
```

When reran with --trace, I get

```
==  SetMirrorRolePermissions: migrating =======================================
rake aborted!
An error has occurred, all later migrations canceled:

undefined method `add_permission!' for nil:NilClass
/var/www/redmine-trunk/vendor/plugins/redmine_git_hosting/db/migrate/20091119162426_set_mirror_role_permissions.rb:7:in `up_without_benchmarks'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:282:in `send'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:282:in `migrate'
/usr/lib/ruby/1.8/benchmark.rb:293:in `measure'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:282:in `migrate'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:365:in `__send__'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:365:in `migrate'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:491
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:567:in `call'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:567:in `ddl_transaction'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:490:in `migrate'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:477:in `each'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:477:in `migrate'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:401:in `up'
/usr/lib/ruby/gems/1.8/gems/activerecord-2.3.11/lib/active_record/migration.rb:383:in `migrate'
/var/www/redmine-trunk/vendor/plugins/engines/lib/engines/plugin/migrator.rb:20:in `migrate_plugin'
/var/www/redmine-trunk/config/../vendor/plugins/engines/lib/engines/plugin.rb:93:in `migrate'
/var/www/redmine-trunk/lib/tasks/migrate_plugins.rake:8
/var/www/redmine-trunk/lib/tasks/migrate_plugins.rake:5:in `each'
/var/www/redmine-trunk/lib/tasks/migrate_plugins.rake:5
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:205:in `call'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:205:in `execute'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:200:in `each'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:200:in `execute'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:158:in `invoke_with_call_chain'
/usr/lib/ruby/1.8/monitor.rb:242:in `synchronize'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:151:in `invoke_with_call_chain'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/task.rb:144:in `invoke'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:112:in `invoke_task'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:90:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:90:in `each'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:90:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:129:in `standard_exception_handling'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:84:in `top_level'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:62:in `run'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:129:in `standard_exception_handling'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/lib/rake/application.rb:59:in `run'
/usr/lib/ruby/gems/1.8/gems/rake-0.9.1/bin/rake:32
/usr/bin/rake:19:in `load'
/usr/bin/rake:19
Tasks: TOP => db:migrate_plugins
```

I have no idea whether this is related, so please tell me whether I should open a new ticket.

@perryk: Where can I set a public key for a user?

**Comment by ericpaulbishop** [comment]
rehmsen: By any chance are you using a locale other than English?  There was a bug I fixed this morning that caused a problem similar to what you are seeing with installations that aren't using an English localization.  It should be fixed now.

**Comment by rehmsen** [comment]
Yes, I am using German locale, but even after pulling and again migrate_plugins I get the same error message.

**Comment by ericpaulbishop** [comment]
If you go into your main redmine directory and run RAILS_ENV=production script/console and then on the command line that comes up type:

I18n.t(:default_role_manager)

What is the output?  Note that second character is a one not a lowercase 'L'

If you get something other than an error, what is the output of this:

Role.find_by_name(I18n.t(:default_role_manager))

**Comment by rehmsen** [comment]
```
>> I18n.t(:default_role_manager)
=> "Manager"
>> Role.find_by_name(I18n.t(:default_role_manager))
=> nil
```

This is the newest commit I pulled.

```
olrehm@h1483674:/var/www/redmine/vendor/plugins/redmine_git_hosting$ git log
commit 0d6f18a2497f8d2dd6f55a754bed1e87b2b0e907
Author: Eric Bishop <ericpaulbishop@gmail.com>
Date:   Tue Sep 6 12:58:16 2011 -0400

    make recursionCheck a class variable, as it should be and perform actual distruction of GitolitePublicKey when we update repos, go back to setting the key as merely inactive in the controller.  Doing it this way means removal of key f

```

Thank you for your help!

**Comment by ericpaulbishop** [comment]
Ok, I just added a begin/rescue block to deal with the case you describe, where you evidently don't have a manager role defined.  It won't kill anything not to have this updated, you'll just have to add mirror permissions manually to the roles you do have.  This should allow the migration to run as expected.

Try again with the latest commit.

**Comment by rehmsen** [comment]
Now it went through - thank you again.

I still have my original issue however: When I remove the repository and recreate it, the one user asscociated with the redmine project (as Developer and Administrator) receives read permissions - but no write permissions. Does that have to do with the Manager role you are talking about? In the permission report, both Developer and Administrator have all permissions listed in the category "Repository".

**Comment by rehmsen** [comment]
Okay, I finally figured it out. Two things could be clarified in the readme:
1. Obviously one has to give the plugin the public key somewhere. It took me a couple of days though to find out how to do this - when I open the user panel there is nowhere the option to do so (at least for me). Instead, I have to go to an empty repository, then click on the _Set public SSH Key_ that I overlooked a hundred times between all the other advice (pushing stuff...) and then suddenly comes an edit user page that has the option to define public keys.
2. If one had a public key for that user already in the keydir (because you used gitolite before, without the plugin), you will have to delete it. Otherwise, gitolite might find it first and find out you have no permissions, and then does not look further.

Thank you again for your help and the great plugin. Now I hope I can finally roll.
